feat(line_numbers): show statuscolumn if available

akinsho opened the pull request:
I believe this will fix #211 without requiring much (barely any) added complexity within this function. Rather than deal with the statuscolumn at all, with this PR users can use the `line_numbers` option to control how the numbers are rendered for each line of the context window i.e. they can re-arrange the padding as needed to align with a custom statusline

<img width="664" alt="image" src="https://user-images.githubusercontent.com/22454918/229198747-c20a1d04-f48e-4644-b6ff-1af250df0831.png">


In the example above in order to match with my statuscolumn which outputs a string that looks like
(note this is not exactly how my statuscolumn is laid out)
```
   1,200  |
   1,201  |
```

I used

```lua
line_numbers = function(lnum, width)
     -- split the padding to surround the number to match the layout of my statuscolumn
	 return ('  %-' .. (width - 4) .. 'd   '):format(lnum)
end,
```

Hopefully this solution allows this plugin to remain agnostic to stuff like the statuscolumn but allows a user to customize the appearance to match their chosen layout.

NOTE: this can easily be a separate option, I just thought it fitting to expand the functionality of this option since it already relates to exactly the same thing

registerGen commented:
Great. But we should also be reminded that this is just a workaround.

akinsho commented:
@registerGen having looked through the code base and also considered how to implement this in a way that does not put too much maintenance burden on the authors/ looking at how this plugin is already structured (a bit) I don't see this as a "workaround". When you say complete fix presumably you imagine that this plugin could automagically get the value of your statuscolumn but at line x that is shown in the context.

From what I can see this plugin *constructs the signcolumn* by looking at what number should be there and what the text offset is.

There is no API to query the value of the statuscolumn at line x AFAICT.

> we need to look into the statusline.

Even if we could do this this is a major pain for the authors to deal with. Even though `nvim_eval_statusline` exists, how will they know what is truly the line number even if you could get access to the statuscolumn *which you cannot*. What if the user puts other numbers in there? or there are signs with numbers? e.g. `|  2  2  |` say the first two is the fold level and the second is the line number...🙃.

Whereas this solution is a 2 or so lines of code and all the user has to do is layout the string the way they have already done for their statuscolumn themselves since they already know what it should look like. Possibly that could be made easier but that's a separate issue.

As always though as this is FOSS if you have some alternative suggestion I'm happy to take those on board this is just my 2 cents/attempt at a minimal solution, since I don't see a simple one but I could easily be missing some alternative

lewis6991 commented:
Honestly I would rather support this properly with no config then add this. If no API exists yet to do this then I'd rather not do it at all.

akinsho commented:
@lewis6991 that's fair enough, it's up to you.

Given that a proper API for this sounds like something that even if it were desirable which you'd know better than me if it would be or not, seems like it would require quite a lot of discussion and possibly a fairly involved implementation. All of which is to say that at best it's a maybe/long way away.

I appreciate not wanting to add this configuration but given the blocker of not having an API for this and that more and more users will start using the statuscolumn once 0.9 is stable. So this will crop up again. Is there some other acceptable middle ground for users currently in this situation? Currently disabling line numbers is possible but then the window just sits beside the statuscolumn but the items in the column relate to the actual lines underneath the window, which I think is fairly confusing since it's easy to think that the line number relates to line in the context or if there was a diagnostic then that there's a bug at that line.

<img width="443" alt="image" src="https://user-images.githubusercontent.com/22454918/229278946-8bf77ce6-1ada-4fe0-bb86-933ae69ec393.png">

Could there be a version of this window that extends the full width at least to cover the statuscolumn so it doesn't look like the signs in it are pointing at the content of the context window

Is there a version of this PR that is palatable e.g. maybe the user doesn't specify a function but just an offset for the line number i.e. at the moment the plugin assumes the line number is always last and so all padding should be before it. Maybe it could allow the user to specify the offset for the number themself and do the padding around it in the plugin?

lewis6991 commented:
Basically, I want a route that allows this config to not exist in the future. If we add it, then there's expectation that it should always be there and work.

> seems like it would require quite a lot of discussion and possibly a fairly involved implementation.

I don't know. It might not. I think it would be reasonable to achieve this via `nvim_eval_statusline` with some provided context.

@luukbvaal: what do you think? Can we get floating windows to emulate `statuscolumn` for windows it's on top of with custom values for the line number?

luukvbaal commented:
Well yeah sounds like `nvim_eval_statusline` may be used to at least parse the `'statuscolumn'` width of the underlying window so that the line number aligns.

Alternatively you might actually copy `'statuscolumn'` to the floating window and make sure the line you're displaying is on the same line number? I.e. fill the buffer with enough empty lines so that `'statuscolumn'` is evaluated with the proper `v:lnum`.

lewis6991 commented:
> Alternatively you might actually copy 'statuscolumn' to the floating window and make sure the line you're displaying is on the same line number? I.e. fill the buffer with enough empty lines so that 'statuscolumn' is evaluated with the proper v:lnum.

That sounds hard and hacky? Could we provide options to `nvim_eval_statusline` to override context?

luukvbaal commented:
Well a limited "context" could be provided by setting `v:lnum` before calling `nvim_eval_statusline()`, which may be enough for simpler `'statuscolumn'`s. Additionally `g:statusline_winid` may be set to the underlying window id. A proper (more complicated) `'statuscolumn'` will use that (through `%!(v:lua.)StatusCol()`) at which point the context may be complete in `nvim_eval_statusline()`.

akinsho commented:
<img width="914" alt="image" src="https://user-images.githubusercontent.com/22454918/229284580-9e10552f-90ba-4d00-a4fa-e495c9f0f63f.png">

Hmm I can get the statuslcolumn text showing with something like
```lua
local function build_lno_str(lnum, width)
  if vim.wo.statuscolumn and #vim.wo.statuscolumn > 0  then
    local key = vim.wo.relativenumber and 'relnum' or 'lnum'
    local prev_value = vim.v[key]
    vim.v[key] = lnum
    local str = api.nvim_eval_statusline(vim.wo.statuscolumn, {maxwidth = 0}).str
    vim.v[key] = prev_value -- does this need to be restored?
    return str
  end
  return string.format('%'..width..'d', lnum)
end
```
Doesn't quite work at the moment since all the line numbers are the same, but just wondering if this is "right"/ an acceptable direction to go in @lewis6991

luukvbaal commented:
> -- does this need to be restored?

Not necessary no, the redraw routines will always set it accordingly.

You should set both `v:relnum` and `v:lnum`, the `'statuscolumn'` might evaluate both.

luukvbaal commented:
If you want to this right we will probably need to add a `use_statuscol` field to `nvim_eval_statusline()` that supplies a non `NULL` `statuscol_T` and fill it's `sattr` and `foldinfo` fields accordingly so that the sign and fold `%s/%C` items can be evaluated properly...

akinsho commented:
Ah I'd forgotten about those (since I personally quite like that they aren't there), and that they wouldn't render correctly probably since were essentially trying to recreate the context for rendering the statuscolumn manually in the example above. Without that will the `%s/%C` columns render as strings or explode is some unforeseen way.

luukvbaal commented:
They will not be drawn, as in evaluated as an empty string which could result in misalignment again.

lewis6991 commented:
Our implementation could pad out the signs to match the window below.

lewis6991 commented:
All in all this sounds fairly reasonable. I'd much rather go this route since: 1) less config, 2) other plugins can benefit.

akinsho commented:
@lewis6991 just to clarify when you say this route, you mean proceeding with implementing what @luukvbaal suggested *without the signs*. Not sure how we'd know how wide that column should be since it could be a variable width and positioned anywhere in the statuscolumn, same goes for the fold column. Or waiting on a change to `nvim_eval_statusline` so signs are populated?

lewis6991 commented:
We should need to provide enough signs in the window and let `nvim_eval_statusline`. Again for folds, we need to figure out how many folds there are in the window.

luukvbaal commented:
> We need to provide enough signs in the window and let nvim_eval_statusline do the rest. Again for folds, we need to figure out how many folds there are in the window and set them in the window.

You shouldn't have to do that if you provide the handle of the underlying window to `nvim_eval_statusline()`.

lewis6991 commented:
Ok, so we just need to fake the line numbers then?

luukvbaal commented:
I think so yeah. 

I considered adapting `nvim_eval_statusline` earlier but it wasn't really an option back then, the refactor in https://github.com/neovim/neovim/pull/22108 made it feasible.

lewis6991 commented:
If it's not too much work, I'm sure @akinsho would really appreciate these improvements. 😉

akinsho commented:
If it was officially supported and the signs & folds were popuplated that'd be great. Since otherwise

> We need to provide enough signs in the window and let nvim_eval_statusline do the rest

I'm not sure what this means/how that would work? the signcolumn can be anywhere between 1-9 right and so presumably when using `%s` in the statuscolumn (which I don't personally) this value could evaluate to a string of length varying from 1-9, but without a proper API it would just be an empty string `""` or would it be `"     "` (size of the column). Similarly for the fold column (it's size is also variable).

Although sounds like it maybe won't be an issue 🤷🏾‍♂️ either way

luukvbaal commented:
@akinsho you could try out this branch: https://github.com/luukvbaal/neovim/tree/evalstc. It seems to work. 

Ironically this might be more efficient than what I'm doing currently in my statuscolumn plugin. For which supporting extmarks was my original motivation for https://github.com/neovim/neovim/pull/22778😅. Although iterating over the signs/marks properly like I'm doing now is more flexible and probably still necessary to support everything I support now.

akinsho commented:
🎉 Thanks for your quick work on that @luukvbaal, I'll give it a proper go in a little bit as I'm not near my computer atm. With `use_statuscol = true` will I still need to set `v:lnum` and `v:relnum` as well as the `statuswinid`, just checking. I don't really mind either way.

luukvbaal commented:
Yes, you will still need to set the variables. I'm not familiar with this plugin but the screenshot seems to show multiple lines. So you would have to increment it to increment `v:lnum` and reevaluate for each line that is drawn.

akinsho commented:
👍🏾 sorry should have clarified with the questions about the variables that I hadn't looked at your commit yet so didn't know if it would take the line number as an argument or something (like I said not necessary was just checking).

Re. the `v:relnum` behaviour the right numbers are actually already calculated by this plugin AFAIK, it's more that when I do `vim.v.relnum = my_lnum` it seems to have no effect. To be clear `vim.v.lnum = num` works fine. But setting `v:relnum` seems not to. Was just wondering if there was some obscure reason I might not have thought of for that. Anyway I'll sit down and test it all properly in a bit and report back.

luukvbaal commented:
> it's more that when I do vim.v.relnum = my_lnum it seems to have no effect

What are you setting it to and what makes you say it has no effect?

akinsho commented:
So I have the following code, and the output beneath it

```lua
  if vim.wo[win].statuscolumn and vim.wo[win].statuscolumn ~= ""  then
    vim.v.lnum = lnum
    local prev_relnum = vim.v.relnum
    vim.v.relnum = relnum
    print(vim.inspect({lnum = lnum, relnum = relnum, vlum = vim.v.lnum, vrlnum = vim.v.relnum, prev_relnum = prev_relnum}))
    vim.g.statusline_winid = win
    local str = api.nvim_eval_statusline(vim.wo[win].statuscolumn, {
      winid = win,
      use_statuscol = true,
      maxwidth = 0,
    }).str
    return str

```
The output is:
```
{
  lnum = 299,
  prev_relnum = 30,
  relnum = 21,
  vlum = 299,
  vrlnum = 30
}
{
  lnum = 200,
  prev_relnum = 29,
  relnum = 120,
  vlum = 200,
  vrlnum = 29
}
{
  lnum = 244,
  prev_relnum = 29,
  relnum = 76,
  vlum = 244,
  vrlnum = 29
}
{
  lnum = 299,
  prev_relnum = 29,
  relnum = 21,
  vlum = 299,
  vrlnum = 29
}

```
Note that `vrlnum` and `relnum` should have the same value as I've just set one to the other just the line before.

luukvbaal commented:
Right, `v:relnum` is read-only. I guess we can lift that restriction.

akinsho commented:
Just done some testing with a build of your commit @luukvbaal and I'm getting a crash using the above code with i.e. the current state of this branch with the `use_statuscol` option and the following statuscolumn `vim.o.statuscolumn = ' %=%l %s %C'`. This reproduces the moment the context window is shown.

I don't have a stack trace nothing was output from the crash (not familiar enough with neovim the project to know how to get one) but it definitely relates to `use_statuscol` since disabling that prevents the crash.

luukvbaal commented:
I admittedly only tested `%s`, perhaps the issue is in `%C`. I can check later.

akinsho commented:
Just tried to narrow the crash down and seems to be anything in the status column other than `%s` so `statuscol='%s'` is fine `statuscol='%s%l` will crash

lewis6991 commented:
@akinsho are you able to get a backtrace when there's a crash? That narrows down things a lot.

luukvbaal commented:
Or simply a repro example without this plugin would be fine, just calling `nvim_eval_status()` directly.

akinsho commented:
The most recent race I got was

```
-------------------------------------
Translated Report (Full Report Below)
-------------------------------------

Process:               nvim [24999]
Path:                  /Users/USER/*/nvim
Identifier:            nvim
Version:               ???
Code Type:             ARM-64 (Native)
Parent Process:        Exited process [24998]
Responsible:           kitty [1255]
User ID:               501

Date/Time:             2023-04-01 16:30:31.1830 +0100
OS Version:            macOS 13.3 (22E252)
Report Version:        12
Anonymous UUID:        B4FD7C53-2291-FECA-2129-083C17DC892A

Sleep/Wake UUID:       D599DFC5-7519-43E5-B0FA-0C5E3CB57433

Time Awake Since Boot: 48000 seconds
Time Since Wake:       1852 seconds

System Integrity Protection: enabled

Crashed Thread:        0  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x0000000000323730
Exception Codes:       0x0000000000000001, 0x0000000000323730

Termination Reason:    Namespace SIGNAL, Code 11 Segmentation fault: 11
Terminating Process:   exc handler [24999]

VM Region Info: 0x323730 is not in any region.  Bytes before following region: 4330899664
      REGION TYPE                    START - END         [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                      102568000-102a5c000    [ 5072K] r-x/r-x SM=COW  ...s/USER/*/nvim

Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	       0x1874dbec4 _platform_strlen + 4
1   libsystem_c.dylib             	       0x18734cf88 __strcat_chk + 52
2   nvim                          	       0x1029469f8 build_stl_str_hl + 5356
3   nvim                          	       0x102699400 nvim_eval_statusline + 1944
4   nvim                          	       0x10265e874 nlua_api_nvim_eval_statusline + 376
5   nvim                          	       0x1025c1414 lj_BC_FUNCC + 44
6   nvim                          	       0x1025d7264 lua_pcall + 228 (lj_api.c:1145)
7   nvim                          	       0x1026aa420 getnextac + 1544
8   nvim                          	       0x1027768e8 do_cmdline + 2280
9   nvim                          	       0x1026a8120 apply_autocmds_group + 2636
10  nvim                          	       0x1026664a8 nvim_exec_autocmds + 1512
11  nvim                          	       0x10264bc90 nlua_api_nvim_exec_autocmds + 592
12  nvim                          	       0x1025c1414 lj_BC_FUNCC + 44
13  nvim                          	       0x1025d7264 lua_pcall + 228 (lj_api.c:1145)
14  nvim                          	       0x102941cd8 state_handle_k_event + 176
15  nvim                          	       0x1028604a8 nv_event + 60
16  nvim                          	       0x1028540dc normal_execute + 9564
17  nvim                          	       0x102941a24 state_enter + 160
18  nvim                          	       0x102646f78 main + 14344
19  dyld                          	       0x187157f28 start + 2236

Thread 1:
0   libsystem_kernel.dylib        	       0x187473710 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x1874b0574 _pthread_cond_wait + 1232
2   nvim                          	       0x10263dff4 uv_cond_wait + 12 (thread.c:911)
3   nvim                          	       0x10262eb50 worker + 112 (threadpool.c:76)
4   libsystem_pthread.dylib       	       0x1874affa8 _pthread_start + 148
5   libsystem_pthread.dylib       	       0x1874aada0 thread_start + 8

Thread 2:
0   libsystem_kernel.dylib        	       0x187473710 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x1874b0574 _pthread_cond_wait + 1232
2   nvim                          	       0x10263dff4 uv_cond_wait + 12 (thread.c:911)
3   nvim                          	       0x10262eb50 worker + 112 (threadpool.c:76)
4   libsystem_pthread.dylib       	       0x1874affa8 _pthread_start + 148
5   libsystem_pthread.dylib       	       0x1874aada0 thread_start + 8

Thread 3:
0   libsystem_kernel.dylib        	       0x187473710 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x1874b0574 _pthread_cond_wait + 1232
2   nvim                          	       0x10263dff4 uv_cond_wait + 12 (thread.c:911)
3   nvim                          	       0x10262eb50 worker + 112 (threadpool.c:76)
4   libsystem_pthread.dylib       	       0x1874affa8 _pthread_start + 148
5   libsystem_pthread.dylib       	       0x1874aada0 thread_start + 8

Thread 4:
0   libsystem_kernel.dylib        	       0x187473710 __psynch_cvwait + 8
1   libsystem_pthread.dylib       	       0x1874b0574 _pthread_cond_wait + 1232
2   nvim                          	       0x10263dff4 uv_cond_wait + 12 (thread.c:911)
3   nvim                          	       0x10262eb50 worker + 112 (threadpool.c:76)
4   libsystem_pthread.dylib       	       0x1874affa8 _pthread_start + 148
5   libsystem_pthread.dylib       	       0x1874aada0 thread_start + 8


Thread 0 crashed with ARM Thread State (64-bit):
    x0: 0x0000000000323734   x1: 0x0000000000323730   x2: 0x000000000000000a   x3: 0x00000001874dbe06
    x4: 0x0000000000000000   x5: 0x000000016d8919f0   x6: 0x0000000000000000   x7: 0x0000000000000000
    x8: 0x000000016d892093   x9: 0x0000000129008200  x10: 0x0000000000323734  x11: 0x000000016d892090
   x12: 0x0000000000000001  x13: 0x0000000000000001  x14: 0x000000016d891ed0  x15: 0x0000000000000000
   x16: 0x00000001874dbec0  x17: 0x00000001e7431940  x18: 0x0000000000000000  x19: 0x000000016d891e8a
   x20: 0x0000000000323734  x21: 0x0000000000000000  x22: 0x0000000129008220  x23: 0x0000000000000046
   x24: 0x0000000000000000  x25: 0x0000000000000002  x26: 0x0000000000000000  x27: 0x0000000102a79760
   x28: 0x0000000102a795f8   fp: 0x000000016d891ce0   lr: 0x000000018734cf88
    sp: 0x000000016d891cb0   pc: 0x00000001874dbec4 cpsr: 0x80001000
   far: 0x0000000000323730  esr: 0x92000006 (Data Abort) byte read Translation fault

Binary Images:
       0x102568000 -        0x102a5bfff nvim (*) <3e8e5a39-1a35-33e5-b772-58ea844c0fbc> /Users/USER/*/nvim
       0x102c70000 -        0x102c7bfff libintl.8.dylib (*) <95dbe1d0-4dd8-3502-abb6-0b08ff7b4e0f> /opt/homebrew/*/libintl.8.dylib
       0x10362c000 -        0x103687fff c.so (*) <029cb822-b5a1-307f-83d2-c52e3308505b> /Users/USER/*/c.so
       0x1874db000 -        0x1874e1ff3 libsystem_platform.dylib (*) <756cd10d-62a0-3283-9e57-cbaa810c95ac> /usr/lib/system/libsystem_platform.dylib
       0x187347000 -        0x1873c5ff7 libsystem_c.dylib (*) <9277aff7-3cc3-30d0-99b7-c62680da95cf> /usr/lib/system/libsystem_c.dylib
       0x187152000 -        0x1871e053b dyld (*) <6f2c2bb8-4bbc-3b64-b927-d3f3193b6295> /usr/lib/dyld
       0x18746f000 -        0x1874a8fe7 libsystem_kernel.dylib (*) <7acbd9bc-d056-310e-858d-81b116cf6d28> /usr/lib/system/libsystem_kernel.dylib
       0x1874a9000 -        0x1874b5fff libsystem_pthread.dylib (*) <b401cfb3-8dfe-32db-92b3-ba8af0f8ca6e> /usr/lib/system/libsystem_pthread.dylib
               0x0 - 0xffffffffffffffff ??? (*) <00000000-0000-0000-0000-000000000000> ???

External Modification Summary:
  Calls made by other processes targeting this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by this process:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0
  Calls made by all processes on this machine:
    task_for_pid: 0
    thread_create: 0
    thread_set_state: 0

VM Region Summary:
ReadOnly portion of Libraries: Total=917.9M resident=0K(0%) swapped_out_or_unallocated=917.9M(100%)
Writable regions: Total=1.3G written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=1.3G(100%)

                                VIRTUAL   REGION 
REGION TYPE                        SIZE    COUNT (non-coalesced) 
===========                     =======  ======= 
Activity Tracing                   256K        1 
Kernel Alloc Once                   32K        1 
MALLOC                           240.2M       30 
MALLOC guard page                   96K        4 
MALLOC_MEDIUM (reserved)         600.0M        5         reserved VM address space (unallocated)
MALLOC_NANO (reserved)           384.0M        1         reserved VM address space (unallocated)
STACK GUARD                       56.1M        5 
Stack                             40.1M        5 
VM_ALLOCATE                       23.3M      187 
__AUTH                             309K       58 
__AUTH_CONST                      3737K      142 
__DATA                            1674K      137 
__DATA_CONST                      4224K      146 
__DATA_DIRTY                       361K       58 
__LINKEDIT                       801.9M        4 
__OBJC_RO                         66.2M        1 
__OBJC_RW                         2011K        1 
__TEXT                           116.0M      153 
dyld private memory                272K        2 
shared memory                       32K        2 
===========                     =======  ======= 
TOTAL                              2.3G      943 
TOTAL, minus reserved VM space     1.3G      943 



-----------
Full Report
-----------

{"app_name":"nvim","timestamp":"2023-04-01 16:30:31.00 +0100","app_version":"","slice_uuid":"3e8e5a39-1a35-33e5-b772-58ea844c0fbc","build_version":"","platform":1,"share_with_app_devs":0,"is_first_party":1,"bug_type":"309","os_version":"macOS 13.3 (22E252)","roots_installed":0,"incident_id":"BB23505E-D99E-48BC-AFD5-2F641E69E489","name":"nvim"}
{
  "uptime" : 48000,
  "procRole" : "Unspecified",
  "version" : 2,
  "userID" : 501,
  "deployVersion" : 210,
  "modelCode" : "MacBookPro18,1",
  "coalitionID" : 827,
  "osVersion" : {
    "train" : "macOS 13.3",
    "build" : "22E252",
    "releaseType" : "User"
  },
  "captureTime" : "2023-04-01 16:30:31.1830 +0100",
  "incident" : "BB23505E-D99E-48BC-AFD5-2F641E69E489",
  "pid" : 24999,
  "translated" : false,
  "cpuType" : "ARM-64",
  "roots_installed" : 0,
  "bug_type" : "309",
  "procLaunch" : "2023-04-01 16:30:30.5739 +0100",
  "procStartAbsTime" : 1160175927486,
  "procExitAbsTime" : 1160190537147,
  "procName" : "nvim",
  "procPath" : "\/Users\/USER\/*\/nvim",
  "parentProc" : "Exited process",
  "parentPid" : 24998,
  "coalitionName" : "net.kovidgoyal.kitty",
  "crashReporterKey" : "B4FD7C53-2291-FECA-2129-083C17DC892A",
  "responsiblePid" : 1255,
  "responsibleProc" : "kitty",
  "codeSigningID" : "nvim",
  "codeSigningTeamID" : "",
  "codeSigningFlags" : 570556929,
  "codeSigningValidationCategory" : 10,
  "codeSigningTrustLevel" : 0,
  "wakeTime" : 1852,
  "sleepWakeUUID" : "D599DFC5-7519-43E5-B0FA-0C5E3CB57433",
  "sip" : "enabled",
  "vmRegionInfo" : "0x323730 is not in any region.  Bytes before following region: 4330899664\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      __TEXT                      102568000-102a5c000    [ 5072K] r-x\/r-x SM=COW  ...s\/USER\/*\/nvim",
  "exception" : {"codes":"0x0000000000000001, 0x0000000000323730","rawCodes":[1,3290928],"type":"EXC_BAD_ACCESS","signal":"SIGSEGV","subtype":"KERN_INVALID_ADDRESS at 0x0000000000323730"},
  "termination" : {"flags":0,"code":11,"namespace":"SIGNAL","indicator":"Segmentation fault: 11","byProc":"exc handler","byPid":24999},
  "vmregioninfo" : "0x323730 is not in any region.  Bytes before following region: 4330899664\n      REGION TYPE                    START - END         [ VSIZE] PRT\/MAX SHRMOD  REGION DETAIL\n      UNUSED SPACE AT START\n--->  \n      __TEXT                      102568000-102a5c000    [ 5072K] r-x\/r-x SM=COW  ...s\/USER\/*\/nvim",
  "extMods" : {"caller":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"system":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"targeted":{"thread_create":0,"thread_set_state":0,"task_for_pid":0},"warnings":0},
  "faultingThread" : 0,
  "threads" : [{"triggered":true,"id":11845085,"threadState":{"x":[{"value":3290932},{"value":3290928},{"value":10},{"value":6564986374},{"value":0},{"value":6132668912},{"value":0},{"value":0},{"value":6132670611},{"value":4982866432},{"value":3290932},{"value":6132670608},{"value":1},{"value":1},{"value":6132670160},{"value":0},{"value":6564986560,"symbolLocation":0,"symbol":"_platform_strlen"},{"value":8174901568},{"value":0},{"value":6132670090},{"value":3290932},{"value":0},{"value":4982866464},{"value":70},{"value":0},{"value":2},{"value":0},{"value":4339504992,"symbolLocation":0,"symbol":"hl_attr_active"},{"value":4339504632,"symbolLocation":0,"symbol":"ns_hl_fast"}],"flavor":"ARM_THREAD_STATE64","lr":{"value":6563352456},"cpsr":{"value":2147487744},"fp":{"value":6132669664},"sp":{"value":6132669616},"esr":{"value":2449473542,"description":"(Data Abort) byte read Translation fault"},"pc":{"value":6564986564,"matchesCrashFrame":1},"far":{"value":3290928}},"queue":"com.apple.main-thread","frames":[{"imageOffset":3780,"symbol":"_platform_strlen","symbolLocation":4,"imageIndex":3},{"imageOffset":24456,"symbol":"__strcat_chk","symbolLocation":52,"imageIndex":4},{"imageOffset":4057592,"symbol":"build_stl_str_hl","symbolLocation":5356,"imageIndex":0},{"imageOffset":1250304,"symbol":"nvim_eval_statusline","symbolLocation":1944,"imageIndex":0},{"imageOffset":1009780,"symbol":"nlua_api_nvim_eval_statusline","symbolLocation":376,"imageIndex":0},{"imageOffset":365588,"symbol":"lj_BC_FUNCC","symbolLocation":44,"imageIndex":0},{"imageOffset":455268,"sourceLine":1145,"sourceFile":"lj_api.c","symbol":"lua_pcall","imageIndex":0,"symbolLocation":228},{"imageOffset":1319968,"symbol":"getnextac","symbolLocation":1544,"imageIndex":0},{"imageOffset":2156776,"symbol":"do_cmdline","symbolLocation":2280,"imageIndex":0},{"imageOffset":1311008,"symbol":"apply_autocmds_group","symbolLocation":2636,"imageIndex":0},{"imageOffset":1041576,"symbol":"nvim_exec_autocmds","symbolLocation":1512,"imageIndex":0},{"imageOffset":933008,"symbol":"nlua_api_nvim_exec_autocmds","symbolLocation":592,"imageIndex":0},{"imageOffset":365588,"symbol":"lj_BC_FUNCC","symbolLocation":44,"imageIndex":0},{"imageOffset":455268,"sourceLine":1145,"sourceFile":"lj_api.c","symbol":"lua_pcall","imageIndex":0,"symbolLocation":228},{"imageOffset":4037848,"symbol":"state_handle_k_event","symbolLocation":176,"imageIndex":0},{"imageOffset":3114152,"symbol":"nv_event","symbolLocation":60,"imageIndex":0},{"imageOffset":3064028,"symbol":"normal_execute","symbolLocation":9564,"imageIndex":0},{"imageOffset":4037156,"symbol":"state_enter","symbolLocation":160,"imageIndex":0},{"imageOffset":913272,"symbol":"main","symbolLocation":14344,"imageIndex":0},{"imageOffset":24360,"symbol":"start","symbolLocation":2236,"imageIndex":5}]},{"id":11845580,"frames":[{"imageOffset":18192,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":6},{"imageOffset":30068,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":7},{"imageOffset":876532,"sourceLine":911,"sourceFile":"thread.c","symbol":"uv_cond_wait","imageIndex":0,"symbolLocation":12},{"imageOffset":813904,"sourceLine":76,"sourceFile":"threadpool.c","symbol":"worker","imageIndex":0,"symbolLocation":112},{"imageOffset":28584,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":7},{"imageOffset":7584,"symbol":"thread_start","symbolLocation":8,"imageIndex":7}]},{"id":11845581,"frames":[{"imageOffset":18192,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":6},{"imageOffset":30068,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":7},{"imageOffset":876532,"sourceLine":911,"sourceFile":"thread.c","symbol":"uv_cond_wait","imageIndex":0,"symbolLocation":12},{"imageOffset":813904,"sourceLine":76,"sourceFile":"threadpool.c","symbol":"worker","imageIndex":0,"symbolLocation":112},{"imageOffset":28584,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":7},{"imageOffset":7584,"symbol":"thread_start","symbolLocation":8,"imageIndex":7}]},{"id":11845582,"frames":[{"imageOffset":18192,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":6},{"imageOffset":30068,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":7},{"imageOffset":876532,"sourceLine":911,"sourceFile":"thread.c","symbol":"uv_cond_wait","imageIndex":0,"symbolLocation":12},{"imageOffset":813904,"sourceLine":76,"sourceFile":"threadpool.c","symbol":"worker","imageIndex":0,"symbolLocation":112},{"imageOffset":28584,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":7},{"imageOffset":7584,"symbol":"thread_start","symbolLocation":8,"imageIndex":7}]},{"id":11845583,"frames":[{"imageOffset":18192,"symbol":"__psynch_cvwait","symbolLocation":8,"imageIndex":6},{"imageOffset":30068,"symbol":"_pthread_cond_wait","symbolLocation":1232,"imageIndex":7},{"imageOffset":876532,"sourceLine":911,"sourceFile":"thread.c","symbol":"uv_cond_wait","imageIndex":0,"symbolLocation":12},{"imageOffset":813904,"sourceLine":76,"sourceFile":"threadpool.c","symbol":"worker","imageIndex":0,"symbolLocation":112},{"imageOffset":28584,"symbol":"_pthread_start","symbolLocation":148,"imageIndex":7},{"imageOffset":7584,"symbol":"thread_start","symbolLocation":8,"imageIndex":7}]}],
  "usedImages" : [
  {
    "source" : "P",
    "arch" : "arm64",
    "base" : 4334190592,
    "size" : 5193728,
    "uuid" : "3e8e5a39-1a35-33e5-b772-58ea844c0fbc",
    "path" : "\/Users\/USER\/*\/nvim",
    "name" : "nvim"
  },
  {
    "source" : "P",
    "arch" : "arm64",
    "base" : 4341563392,
    "size" : 49152,
    "uuid" : "95dbe1d0-4dd8-3502-abb6-0b08ff7b4e0f",
    "path" : "\/opt\/homebrew\/*\/libintl.8.dylib",
    "name" : "libintl.8.dylib"
  },
  {
    "source" : "P",
    "arch" : "arm64",
    "base" : 4351770624,
    "size" : 376832,
    "uuid" : "029cb822-b5a1-307f-83d2-c52e3308505b",
    "path" : "\/Users\/USER\/*\/c.so",
    "name" : "c.so"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6564982784,
    "size" : 28660,
    "uuid" : "756cd10d-62a0-3283-9e57-cbaa810c95ac",
    "path" : "\/usr\/lib\/system\/libsystem_platform.dylib",
    "name" : "libsystem_platform.dylib"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6563328000,
    "size" : 520184,
    "uuid" : "9277aff7-3cc3-30d0-99b7-c62680da95cf",
    "path" : "\/usr\/lib\/system\/libsystem_c.dylib",
    "name" : "libsystem_c.dylib"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6561275904,
    "size" : 582972,
    "uuid" : "6f2c2bb8-4bbc-3b64-b927-d3f3193b6295",
    "path" : "\/usr\/lib\/dyld",
    "name" : "dyld"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6564540416,
    "size" : 237544,
    "uuid" : "7acbd9bc-d056-310e-858d-81b116cf6d28",
    "path" : "\/usr\/lib\/system\/libsystem_kernel.dylib",
    "name" : "libsystem_kernel.dylib"
  },
  {
    "source" : "P",
    "arch" : "arm64e",
    "base" : 6564777984,
    "size" : 53248,
    "uuid" : "b401cfb3-8dfe-32db-92b3-ba8af0f8ca6e",
    "path" : "\/usr\/lib\/system\/libsystem_pthread.dylib",
    "name" : "libsystem_pthread.dylib"
  },
  {
    "size" : 0,
    "source" : "A",
    "base" : 0,
    "uuid" : "00000000-0000-0000-0000-000000000000"
  }
],
  "sharedCache" : {
  "base" : 6560612352,
  "size" : 3539501056,
  "uuid" : "581a6955-d455-3735-973b-5233bd5dd16e"
},
  "vmSummary" : "ReadOnly portion of Libraries: Total=917.9M resident=0K(0%) swapped_out_or_unallocated=917.9M(100%)\nWritable regions: Total=1.3G written=0K(0%) resident=0K(0%) swapped_out=0K(0%) unallocated=1.3G(100%)\n\n                                VIRTUAL   REGION \nREGION TYPE                        SIZE    COUNT (non-coalesced) \n===========                     =======  ======= \nActivity Tracing                   256K        1 \nKernel Alloc Once                   32K        1 \nMALLOC                           240.2M       30 \nMALLOC guard page                   96K        4 \nMALLOC_MEDIUM (reserved)         600.0M        5         reserved VM address space (unallocated)\nMALLOC_NANO (reserved)           384.0M        1         reserved VM address space (unallocated)\nSTACK GUARD                       56.1M        5 \nStack                             40.1M        5 \nVM_ALLOCATE                       23.3M      187 \n__AUTH                             309K       58 \n__AUTH_CONST                      3737K      142 \n__DATA                            1674K      137 \n__DATA_CONST                      4224K      146 \n__DATA_DIRTY                       361K       58 \n__LINKEDIT                       801.9M        4 \n__OBJC_RO                         66.2M        1 \n__OBJC_RW                         2011K        1 \n__TEXT                           116.0M      153 \ndyld private memory                272K        2 \nshared memory                       32K        2 \n===========                     =======  ======= \nTOTAL                              2.3G      943 \nTOTAL, minus reserved VM space     1.3G      943 \n",
  "legacyInfo" : {
  "threadTriggered" : {
    "queue" : "com.apple.main-thread"
  }
},
  "logWritingSignature" : "45aab1b94f4f2dc9d27536d367af6aea32fe0da8",
  "trialInfo" : {
  "rollouts" : [
    {
      "rolloutId" : "62fe74515312cd4599bd3c80",
      "factorPackIds" : {
        "MYRIAD_BOOSTS" : "62fe74805312cd4599bd3c81"
      },
      "deploymentId" : 240000006
    },
    {
      "rolloutId" : "60186475825c62000ccf5450",
      "factorPackIds" : {

      },
      "deploymentId" : 240000026
    }
  ],
  "experiments" : [
    {
      "treatmentId" : "c28e4ee6-1b08-4f90-8e05-2809e78310a3",
      "experimentId" : "6317d2003d24842ff850182a",
      "deploymentId" : 400000013
    }
  ]
}
}
```

akinsho commented:
I'm also able to reproduce it with the following minimal init
```lua
vim.o.signcolumn = 'yes:1'
vim.o.number = true
vim.o.relativenumber = true
vim.o.statuscolumn = ' %=%l %s %C'

vim.schedule(function()
  local str = vim.api.nvim_eval_statusline(vim.o.statuscolumn, {
    winid = vim.api.nvim_get_current_win(),
    use_statuscol = true,
  }).str
  vim.notify('Str: ' .. str)
end)
```

luukvbaal commented:
Sorry, I cant reproduce the crash. I had experienced crashes myself which I already fixed, I don't quite remember if I force pushed any change after posting the branch here. Are you on the latest commit?

akinsho commented:
🤔 I've just pulled from your branch and it seems to have now updated things, I've removed the build dir and the deps and rebuilt it and it still consistently crashes for me with 
`VIMRUNTIME=runtime ./build/bin/nvim -u ~/minimal.lua ./src/spellfile.c`.

Just to be clear when you test it you need to ensure:
1. You have the `signcolumn`, `relativenumber` and `number` set, (all covered by the minimal config)
2. You need to actually be in a non empty file so there are some actual results to return.

luukvbaal commented:
I'll make a proper PR tomorrow and look into the crash again.

luukvbaal commented on lua/treesitter-context.lua:
```diff
@@ -529,10 +529,25 @@ local function highlight_contexts(bufnr, ctx_bufnr, contexts)
+local function build_lno_str(win, lnum, relnum, width)
+  if vim.wo[win].statuscolumn and vim.wo[win].statuscolumn ~= ""  then
+    vim.v.lnum = lnum
+    vim.v.relnum = relnum -- TODO: this variable is currently readonly
```
These can both be removed, both are set accordingly depending on `v:lnum` in `nvim_eval_statusline()`. (which happens to not be read-only so we don't have to change anything there in https://github.com/neovim/neovim/pull/22859).

luukvbaal commented on lua/treesitter-context.lua:
```diff
@@ -529,10 +529,25 @@ local function highlight_contexts(bufnr, ctx_bufnr, contexts)
+    local ok, data = pcall(api.nvim_eval_statusline,vim.wo[win].statuscolumn, {
+      winid = win,
+      use_statuscol = true,
+      maxwidth = 0,
```
This is also calculated accordingly (`win_col_off()`) and should be removed.

> the statuscolumn is initially missing some content, as the numbers are flush with the the rest of the content, but on moving my cursor the first time they get populated.

I'm not sure if that will fix this issue. Is `'statuscolumn'` set before this plugin loads in your config?

akinsho replied:
@luukvbaal I believe you are right and it is a timing issue. AFAICT my statuscolumn should be coming in way before the context window, as that is loaded as a `plugin/statuscolumn`, I even tried delaying it to `CursorHold` but still seem that for whatever reason the context window starts first. Anyway this isn't relevant to your PR I think which is working nicely 👍🏾

akinsho replied:
I guess it' probably because it's an nvim-treesitter module and so it is required quite early in the startup process and so actually runs before `plugin/statuscolumn` 🤷🏾‍♂️ (just thinking aloud not relate to your work @luukvbaal )

luukvbaal commented:
@akinsho FYI `opts->use_statuscol` was renamed to `use_statuscol_lnum` and should be set to the line number. No need to modify `v:lnum` directly.

akinsho commented:
Thanks for the heads up @luukvbaal and for all your work on that and the previous PR 🙏🏾
